### PR TITLE
crypto/sm2/sm2_pmeth.c: add sm2 default userid

### PR DESCRIPTION
**Pull request by moehuster**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/sm2/sm2_pmeth.c [hunk 1]
```diff
@@ -32,12 +32,24 @@ typedef struct {
...
+    uint8_t *id = SM2_DEFAULT_USERID;
+    size_t id_len = strlen(id);
+    uint8_t *tmp_id;
 
     if ((smctx = OPENSSL_zalloc(sizeof(*smctx))) == NULL) {
         SM2err(SM2_F_PKEY_SM2_INIT, ERR_R_MALLOC_FAILURE);
         return 0;
     }
 
+    tmp_id = OPENSSL_malloc(id_len);
```
Comment (richsalz) [inline crypto/sm2/sm2_pmeth.c]: replace lines 44-49 with OPENSSL_strdup ?

Comment (moehuster) [reply crypto/sm2/sm2_pmeth.c]: OPENSSL_strdup need more than once strlen invoketion

Comment (richsalz) [reply crypto/sm2/sm2_pmeth.c]: Optimizing a strlen compared with code clarity is the wrong thing to do.

Comment (moehuster) [reply crypto/sm2/sm2_pmeth.c]: got it!

**Comment by levitte** [comment]
This should definitely be cherry-picked up to master as well, so I've labelled this for both master and 1.1.1.  *To anyone who merges:* it should be absolutely fine to start with applying this on 1.1.1 and cherry-pick it up to master, but the reverse should be harmless as well.

**Comment by levitte** [comment]
It looks like this triggered a bug somewhere, see https://travis-ci.org/openssl/openssl/jobs/605862230

More precisely, there's a memory leak.  And just FYI, it's *not* this PR's fault as far as I can see, so I think we've a case of sloppy cleaning somewhere else.

```
    # [05:46:59]  3748 file=crypto/sm2/sm2_pmeth.c, line=43, thread=47963119216704, number=17, address=0x13fe600
    # [05:46:59]  4537 file=crypto/sm2/sm2_pmeth.c, line=43, thread=47963119216704, number=17, address=0x13fe640
    # [05:46:59]  3555 file=crypto/sm2/sm2_pmeth.c, line=43, thread=47963119216704, number=17, address=0x1401100
    # 51 bytes leaked in 3 chunks
/home/travis/build/openssl/openssl/util/shlib_wrap.sh /home/travis/build/openssl/openssl/test/evp_extra_test => 1
not ok 1 - running evp_extra_test
```

**Comment by mattcaswell** [comment]
@InfoHunter - please take a look at this.

**Comment by moehuster** [comment]
> @InfoHunter - please take a look at this.

sorry for commit nothing, just ignore it

**Comment by InfoHunter** [comment]
I am not sure yet if we need to set the default ID to SM2 - discussions welcome ;-)

**Comment by moehuster** [comment]
We use the default ID in most cases. User defined ids only appear in TEST scripts.

**Comment by InfoHunter** [comment]
> We use the default ID in most cases. User defined ids only appear in TEST scripts.

I was not saying the default ID should not be used. I was just explaining why OpenSSL doesn't set the ID by default. And there is indeed a reason about this:

For what OpenSSL has discussed before, we decided at that time it's the application's responsibility to set the default SM2 ID instead of OpenSSL's. So I don't know it's worth to change this consideration or not...

**Comment by moehuster** [comment]
> It looks like this triggered a bug somewhere, see https://travis-ci.org/openssl/openssl/jobs/605862230
> 
> More precisely, there's a memory leak. And just FYI, it's _not_ this PR's fault as far as I can see, so I think we've a case of sloppy cleaning somewhere else.
> 
> ```
>     # [05:46:59]  3748 file=crypto/sm2/sm2_pmeth.c, line=43, thread=47963119216704, number=17, address=0x13fe600
>     # [05:46:59]  4537 file=crypto/sm2/sm2_pmeth.c, line=43, thread=47963119216704, number=17, address=0x13fe640
>     # [05:46:59]  3555 file=crypto/sm2/sm2_pmeth.c, line=43, thread=47963119216704, number=17, address=0x1401100
>     # 51 bytes leaked in 3 chunks
> /home/travis/build/openssl/openssl/util/shlib_wrap.sh /home/travis/build/openssl/openssl/test/evp_extra_test => 1
> not ok 1 - running evp_extra_test
> ```

**Comment by t8m** [comment]
This would need to be submitted against the master branch to be considered. This is not a bug fix to be done on a released branch.

**Pull request closed without merging**